Userland/Applications/SoundPlayer/SoundPlayerWidgetAdvancedView.cpp

```diff
@@ -249,85 +145,96 @@ void SoundPlayerWidgetAdvancedView::keydown_event(GUI::KeyEvent& event)
+        break;
+    case PlayState::Playing:
+        m_play_button->set_icon(*m_pause_icon);
+        m_play_button->set_enabled(true);
+        m_stop_button->set_enabled(true);
+        m_playback_progress_slider->set_enabled(true);
+        break;
+    case PlayState::Paused:
+        m_play_button->set_icon(*m_play_icon);
+        m_play_button->set_enabled(true);
+        m_stop_button->set_enabled(true);
+        m_playback_progress_slider->set_enabled(true);
+        break;
+    case PlayState::Stopped:
+        m_play_button->set_icon(*m_play_icon);
+        m_play_button->set_enabled(true);
+        m_stop_button->set_enabled(false);
+        m_playback_progress_slider->set_enabled(true);
+        break;
+    }
 }
 
-void SoundPlayerWidgetAdvancedView::read_playlist(StringView path)
+void SoundPlayerWidgetAdvancedView::loop_mode_changed(Player::LoopMode)
 {
-    auto parser = M3UParser::from_file(path);
-    auto items = parser->parse(true);
-    VERIFY(items->size() > 0);
-    try_fill_missing_info(*items, path);
-    for (auto& item : *items)
-        m_playlist_model->items().append(item);
-    set_playlist_visible(true);
-    m_playlist_model->invalidate();
+}
 
-    open_file(items->at(0).path);
+void SoundPlayerWidgetAdvancedView::time_elapsed(int seconds)
+{
+    m_timestamp_label->set_text(String::formatted("Elapsed: {:02}:{:02}:{:02}", seconds / 3600, seconds / 60, seconds % 60));
+}
 
-    if (items->size() > 1) {
-        m_back_button->set_enabled(true);
-        m_next_button->set_enabled(true);
-    } else {
-        m_back_button->set_enabled(false);
-        m_next_button->set_enabled(false);
-    }
+void SoundPlayerWidgetAdvancedView::file_name_changed(StringView name)
+{
+    m_window.set_title(String::formatted("{} - Sound Player", name));
 }
 
-void SoundPlayerWidgetAdvancedView::set_playlist_visible(bool visible)
+void SoundPlayerWidgetAdvancedView::total_samples_changed(int total_samples)
 {
-    if (visible) {
-        if (!m_playlist_widget->parent()) {
-            m_player_view->parent_widget()->add_child(*m_playlist_widget);
-        }
-    } else {
-        m_playlist_widget->remove_from_parent();
-        m_player_view->set_max_width(window()->width());
-    }
+    m_playback_progress_slider->set_max(total_samples);
+    m_playback_progress_slider->set_page_step(total_samples / 10);
 }
 
-void SoundPlayerWidgetAdvancedView::try_fill_missing_info(Vector<M3UEntry>& entries, StringView playlist_p)
+void SoundPlayerWidgetAdvancedView::sound_buffer_played(RefPtr<Audio::Buffer> buffer, int sample_rate, int samples_played)
 {
-    LexicalPath playlist_path(playlist_p);
-    Vector<M3UEntry*> to_delete;
-    for (auto& entry : entries) {
-        if (!LexicalPath(entry.path).is_absolute()) {
-            entry.path = String::formatted("{}/{}", playlist_path.dirname(), entry.path);
-        }
+    auto visualization = dynamic_cast<Visualization*>(m_visualization.ptr());
```

Comment:
Why do you extract the pointer from `m_visualization`, that seems very pointless (lol) given that you don't pass it to anyone. Let's not do this.

Reply:
Just moving some things around. I don't like using RTTI either, so this is something that I'll eventually fix. But one thing at a time.